cmd/pullrequest-init/github_test.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/google/go-github/github"
+	"go.uber.org/zap"
+)
+
+func newClient(gh *FakeGitHub) (*github.Client, func()) {
+	s := httptest.NewServer(gh)
+	client := github.NewClient(s.Client())
+	u, err := url.Parse(fmt.Sprintf("%s/", s.URL))
+	if err != nil {
+		panic(err)
+	}
+	client.BaseURL = u
+	return client, s.Close
+}
+
+func TestGitHub(t *testing.T) {
+	ctx := context.Background()
+	gh := NewFakeGitHub()
+	client, close := newClient(gh)
+	defer close()
+
+	owner := "foo"
+	repo := "bar"
+	prNum := 1
+
+	pr := &github.PullRequest{
+		ID:      github.Int64(int64(prNum)),
+		HTMLURL: github.String(fmt.Sprintf("https://github.com/%s/%s/pull/%d", owner, repo, prNum)),
+		Base: &github.PullRequestBranch{
+			User: &github.User{
+				Login: github.String(owner),
+			},
+			Repo: &github.Repository{
+				Name:     github.String(repo),
+				CloneURL: github.String(fmt.Sprintf("https://github.com/%s/%s", owner, repo)),
+			},
+			Ref: github.String("master"),
+			SHA: github.String("1"),
+		},
+		Head: &github.PullRequestBranch{
+			User: &github.User{
+				Login: github.String("baz"),
+			},
+			Repo: &github.Repository{
+				Name:     github.String(repo),
+				CloneURL: github.String(fmt.Sprintf("https://github.com/baz/%s", repo)),
+			},
+			Ref: github.String("feature"),
+			SHA: github.String("2"),
+		},
+	}
+	gh.AddPullRequest(pr)
+	comment := &github.IssueComment{
+		ID:   github.Int64(1),
+		Body: github.String("hello world!"),
+	}
+	gh.AddComment(owner, repo, int64(prNum), comment)
+
+	h, err := NewGitHubHandler(ctx, zap.NewNop().Sugar(), pr.GetHTMLURL())
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Override GitHub client to point to local test server.
+	h.Client = client
+
+	dir := os.TempDir()
+	if err := h.Download(ctx, dir); err != nil {
+		t.Fatal(err)
+	}
+
+	prPath := filepath.Join(dir, "pr.json")
+	rawPRPath := filepath.Join(dir, "github/pr.json")
+	rawCommentPath := filepath.Join(dir, "github/comments/1.json")
+
+	wantPR := &PullRequest{
+		Type: "github",
+		ID:   int64(prNum),
+		Head: &GitReference{
+			Repo:   pr.GetHead().GetRepo().GetCloneURL(),
+			Branch: pr.GetHead().GetRef(),
+			SHA:    pr.GetHead().GetSHA(),
+		},
+		Base: &GitReference{
+			Repo:   pr.GetBase().GetRepo().GetCloneURL(),
+			Branch: pr.GetBase().GetRef(),
+			SHA:    pr.GetBase().GetSHA(),
+		},
+		Comments: []*Comment{{
+			ID:     comment.GetID(),
+			Author: comment.GetUser().GetLogin(),
+			Text:   comment.GetBody(),
+			Raw:    rawCommentPath,
+		}},
+		Labels: []*Label{},
+		Raw:    rawPRPath,
+	}
+
+	gotPR := new(PullRequest)
+	diffFile(t, prPath, wantPR, gotPR)
+	diffFile(t, rawPRPath, pr, new(github.PullRequest))
+	if rawPRPath != gotPR.Raw {
+		t.Errorf("Raw PR path: want [%s], got [%s]", rawPRPath, gotPR.Raw)
+	}
+	diffFile(t, rawCommentPath, comment, new(github.IssueComment))
+	if rawCommentPath != gotPR.Comments[0].Raw {
+		t.Errorf("Raw PR path: want [%s], got [%s]", rawCommentPath, gotPR.Comments[0].Raw)
+	}
+
+	// Upload
```

Comment:
maybe 2 different tests, one for download, one for upload?

Reply:
> I originally wanted to do this, but I found that the easiest way to test Upload was to call Download, modify the result of download, then upload. Since a lot of the setup (i.e. populating the fake GitHub server) ends up being duplicative, it seemed simpler to test them together.

The setup for upload could be "hard-coded" file(s) instead of doing Download right ? (that said it might be as long of a setup as calling Download :sweat_smile: )

Reply:
Done.